File: implementations/typescript/src/task.js
import { ErrorCode, errorPayload } from "./errors.js";

const TaskState = Object.freeze({
  SUBMITTED: "submitted",
  ACCEPTED: "accepted",
  STARTED: "started",
  BLOCKED: "blocked",
  PROGRESS: "progress",
  OUTPUT: "output",
  COMPLETED: "completed",
  FAILED: "failed",
  CANCELLED: "cancelled",
  TIMED_OUT: "timed_out"
});

const terminalStates = new Set([
  TaskState.COMPLETED,
  TaskState.FAILED,
  TaskState.CANCELLED,
  TaskState.TIMED_OUT
]);

const allowedTransitions = {
  [TaskState.SUBMITTED]: new Set([TaskState.ACCEPTED, TaskState.FAILED, TaskState.CANCELLED, TaskState.TIMED_OUT]),
  [TaskState.ACCEPTED]: new Set([TaskState.STARTED, TaskState.FAILED, TaskState.CANCELLED, TaskState.TIMED_OUT]),
  [TaskState.STARTED]: new Set([TaskState.PROGRESS, TaskState.OUTPUT, TaskState.BLOCKED, TaskState.COMPLETED, TaskState.FAILED, TaskState.CANCELLED, TaskState.TIMED_OUT]),
  [TaskState.BLOCKED]: new Set([TaskState.STARTED, TaskState.PROGRESS, TaskState.FAILED, TaskState.CANCELLED, TaskState.TIMED_OUT]),
  [TaskState.PROGRESS]: new Set([TaskState.PROGRESS, TaskState.OUTPUT, TaskState.BLOCKED, TaskState.COMPLETED, TaskState.FAILED, TaskState.CANCELLED, TaskState.TIMED_OUT]),
  [TaskState.OUTPUT]: new Set([TaskState.PROGRESS, TaskState.OUTPUT, TaskState.BLOCKED, TaskState.COMPLETED, TaskState.FAILED, TaskState.CANCELLED, TaskState.TIMED_OUT])
};

function nowISO() {
  return new Date().toISOString();
}

export class TaskTracker {
  constructor(data = {}) {
    this.id = data.id ?? data.task_id ?? `task_${Date.now().toString(36)}`;
    this.description = data.description ?? data.payload?.description ?? "";
    this.state = TaskState.SUBMITTED;
    this.source = data.source ?? "unknown";
    this.sessionId = data.session_id ?? null;
    this.conversationId = data.conversation_id ?? null;
    this.createdAt = data.created_at ?? nowISO();
    this.updatedAt = this.createdAt;
    this._eventId = 0;
  }

  nextEventId() {
    return `evt_task_${String(++this._eventId).padStart(6, "0")}`;
  }

  transition(eventType, payload = {}) {
    const nextState = eventTypeToState(eventType);
    if (!nextState) {
      throw new Error(`unknown task event type: ${eventType}`);
    }

    if (nextState !== this.state && !allowedTransitions[this.state]?.has(nextState)) {
      throw new Error(
        `illegal task transition: ${this.state} -> ${nextState} for task ${this.id}`
      );
    }

    this.state = nextState;
    this.updatedAt = nowISO();

    const event = {
      aep_version: "0.1",
      id: this.nextEventId(),
      type: eventType,
      source: this.source,
      session_id: this.sessionId,
      conversation_id: this.conversationId,
      task_id: this.id,
      created_at: this.updatedAt,
      payload: {
        task_id: this.id,
        state: this.state,
        ...payload
      }
    };

    if (terminalStates.has(this.state) && !("result" in event.payload)) {
      event.payload.result = this.state;
    }

    return event;
  }

  submitted() { return this.transition("task.submitted"); }
  accepted() { return this.transition("task.accepted"); }
  started() { return this.transition("task.started"); }
  blocked(reason) { return this.transition("task.blocked", reason ? { reason } : {}); }
  progress(data) { return this.transition("task.progress", data); }
  output(data) { return this.transition("task.output", data); }
  completed(result) { return this.transition("task.completed", result ? { result } : {}); }
  failed(code, message, details) {
    return this.transition("task.failed", {
      error: errorPayload(code, message, { details })
    });
  }
  cancelled(reason) {
    return this.transition("task.cancelled", reason ? { reason } : {});
  }
  timedOut() {
    return this.transition("task.timed_out", {
      error: errorPayload(ErrorCode.TASK_TIMEOUT, `task ${this.id} timed out`, { retryable: true })
    });
  }

  isTerminal() {
    return terminalStates.has(this.state);
  }

  isActive() {
    return !this.isTerminal() && this.state !== TaskState.SUBMITTED;
  }
}

function eventTypeToState(type) {
  const map = {
    "task.submitted": TaskState.SUBMITTED,
    "task.accepted": TaskState.ACCEPTED,
    "task.started": TaskState.STARTED,
    "task.blocked": TaskState.BLOCKED,
    "task.progress": TaskState.PROGRESS,
    "task.output": TaskState.OUTPUT,
    "task.completed": TaskState.COMPLETED,
    "task.failed": TaskState.FAILED,
    "task.cancelled": TaskState.CANCELLED,
    "task.timed_out": TaskState.TIMED_OUT
  };
  return map[type] ?? null;
}

File: implementations/go/task/task.go
package task

import (
	"fmt"
	"time"
)

type State string

const (
	Submitted State = "submitted"
	Accepted  State = "accepted"
	Started   State = "started"
	Progress  State = "progress"
	Blocked   State = "blocked"
	Output    State = "output"
	Completed State = "completed"
	Failed    State = "failed"
	Cancelled State = "cancelled"
	TimedOut  State = "timed_out"
)

const TaskTimeout = "task_timeout"

var eventToState = map[string]State{
	"task.submitted": Submitted,
	"task.accepted":  Accepted,
	"task.started":   Started,
	"task.progress":  Progress,
	"task.blocked":   Blocked,
	"task.output":    Output,
	"task.completed": Completed,
	"task.failed":    Failed,
	"task.cancelled": Cancelled,
	"task.timed_out": TimedOut,
}

var terminalStates = map[State]bool{
	Completed: true,
	Failed:    true,
	Cancelled: true,
	TimedOut:  true,
}

var transitions = map[State]map[State]bool{
	Submitted: {Accepted: true, Failed: true, Cancelled: true, TimedOut: true},
	Accepted:  {Started: true, Failed: true, Cancelled: true, TimedOut: true},
	Started:   {Progress: true, Output: true, Blocked: true, Completed: true, Failed: true, Cancelled: true, TimedOut: true},
	Blocked:   {Started: true, Progress: true, Failed: true, Cancelled: true, TimedOut: true},
	Progress:  {Progress: true, Output: true, Blocked: true, Completed: true, Failed: true, Cancelled: true, TimedOut: true},
	Output:    {Progress: true, Output: true, Blocked: true, Completed: true, Failed: true, Cancelled: true, TimedOut: true},
}

func errorPayload(code, message string, retryable bool) map[string]any {
	return map[string]any{
		"code":      code,
		"message":   message,
		"retryable": retryable,
		"details":   map[string]any{},
	}
}

type Tracker struct {
	ID          string
	State       State
	Source      string
	Description string
	eventID     int
}

func NewTracker(id, source, description string) *Tracker {
	return &Tracker{
		ID:          id,
		State:       Submitted,
		Source:      source,
		Description: description,
	}
}

func (t *Tracker) Accept() map[string]any {
	return t.Transition("task.accepted", nil)
}

func (t *Tracker) Accepted() map[string]any {
	return t.Transition("task.accepted", nil)
}

func (t *Tracker) Started() map[string]any {
	return t.Transition("task.started", nil)
}

func (t *Tracker) Progress(payload map[string]any) map[string]any {
	return t.Transition("task.progress", payload)
}

func (t *Tracker) Completed(result map[string]any) map[string]any {
	return t.Transition("task.completed", result)
}

func (t *Tracker) Failed(code, message string) map[string]any {
	return t.Transition("task.failed", map[string]any{
		"error": errorPayload(code, message, false),
	})
}

func (t *Tracker) IsTerminal() bool {
	return terminalStates[t.State]
}

func (t *Tracker) Transition(eventType string, payload map[string]any) map[string]any {
	nextState, ok := eventToState[eventType]
	if !ok {
		return nil
	}

	if nextState != t.State {
		allowed, hasAllowed := transitions[t.State]
		if !hasAllowed || !allowed[nextState] {
			return nil
		}
	}

	t.State = nextState

	result := make(map[string]any, len(payload)+2)
	if payload != nil {
		for k, v := range payload {
			result[k] = v
		}
	}
	result["task_id"] = t.ID
	result["state"] = string(t.State)

	if terminalStates[t.State] {
		if _, ok := result["result"]; !ok {
			result["result"] = string(t.State)
		}
	}

	t.eventID++
	return map[string]any{
		"spec_version": "0.2",
		"id":           fmt.Sprintf("evt_task_%06d", t.eventID),
		"type":         eventType,
		"source":       t.Source,
		"task_id":      t.ID,
		"created_at":   time.Now().UTC().Format(time.RFC3339),
		"payload":      result,
	}
}
